rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+## Summary
+[summary]: #summary
+
+This RFC proposes an extension to our use of lookup tables using the PLOOKUP
+multiset inclusion argument, so that values within lookup tables can be chosen
+after the constraint system for a circuit has been fixed.
+
+## Motivation
+[motivation]: #motivation
+
+This extension should provide us with
+* array-style lookups (`arr[i]`, where `arr` and `i` are formed of values in
+  the proof witness).
+* the ability to load 'bytecode' using a proof-independent commitment.
+* the ability to load tables of arbitrary data for use in the circuit, allowing
+  e.g. the full data of a HTTPS transcript to be entered into a proof without
+  using some in-circuit commitment.
+* the ability to expose tables of arbitrary in-circuit data, for re-use by
+  other proofs.
+
+These goals support 5 major use-cases:
+* allow the verifier circuit to support 'custom' or user-specified gates
+  - using array indexing, the use of constants and commitments/evaluations can
+    be determined by the verifier index, instead of being fixed by the
+    permutation argument as currently.
+* allow random-access lookups in arrays as a base-primitive for user programs
+* allow circuits to use low-cost branching and other familiar programming
+  primitives
+  - this depends on the development of a bytecode interpreter on top of the
+    lookup-table primitives.
```

Comment:
It's interesting to think how this could link-up with what Anais is doing.  I wonder if there are any design considerations that could come out of her work that could impact the design here.

Reply:
Added a clarifying comment.